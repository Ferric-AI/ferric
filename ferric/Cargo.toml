[package]
name = "ferric"
version = "0.1.2"
edition = "2021"
authors = ["Nimar Arora <nimar.arora@gmail.com>"]
license = "MIT OR Apache-2.0"
description = "A Probablistic Programming Language with a declarative syntax for random variables."
readme = "../README.md"
repository = "https://github.com/ferric-ai/ferric"
homepage = "https://ferric.ai"
keywords = ["probabilistic", "programming", "generative", "bayesian", "mcmc"]
categories = ["mathematics", "science"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
ferric-macros = {path="../ferric-macros", version = "0.1.0"}
rand = "0.8"
rand_distr = "0.4"

[dev-dependencies]
ferric = {path="../ferric"}
trybuild = "1.0"
