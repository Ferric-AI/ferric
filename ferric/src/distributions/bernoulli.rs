// Copyright 2022 The Ferric AI Project Developers

use crate::distributions::Distribution;
use rand::Rng;

pub struct Bernoulli {
    /// Probability of success.
    p: f64,
}

impl Bernoulli {
    pub fn new(p: f64) -> Result<Bernoulli, String> {
        if !(0f64..=1f64).contains(&p) {
            Err(format! {"Bernoulli: illegal probability `{}` should be between 0 and 1", p})
        } else {
            Ok(Bernoulli { p })
        }
    }
}

impl<R: Rng + ?Sized> Distribution<R> for Bernoulli {
    type Domain = bool;
    fn sample(&self, rng: &mut R) -> bool {
        let val: f64 = rng.gen();
        val < self.p
    }
}

impl std::fmt::Display for Bernoulli {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Bernoulli {{ p = {} }}", self.p)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::thread_rng;

    #[test]
    fn bernoulli_sample() {
        let mut rng = thread_rng();
        let dist = Bernoulli::new(0.1).unwrap();
        let mut succ = 0;
        let trials = 10000;
        for _ in 0..trials {
            if dist.sample(&mut rng) {
                succ += 1;
            }
        }
        let mean = (succ as f64) / (trials as f64);
        assert!((mean - 0.1).abs() < 0.01);
        // both of the following extreme distributios should be allowed
        let _dist2 = Bernoulli::new(0.0).unwrap();
        let _dist3 = Bernoulli::new(1.0).unwrap();
    }

    #[test]
    #[should_panic]
    fn bernoulli_too_low() {
        let _dist = Bernoulli::new(-0.01).unwrap();
    }

    #[test]
    #[should_panic]
    fn bernoulli_too_high() {
        let _dist = Bernoulli::new(1.01).unwrap();
    }
}
