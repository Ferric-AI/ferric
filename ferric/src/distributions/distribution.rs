// Copyright 2022 The Ferric AI Project Developers

use rand::rngs::ThreadRng;
use rand::Rng;

/// A Distribution<Domain=...> defined over a specific domain can be used to generate
/// random values over that domain.
/// For example, the Bernoulli distribution implements Distribution<Domain=bool> and
/// it can be used to generate random booleans.
pub trait Distribution<R = ThreadRng>
where
    R: Rng + ?Sized,
{
    type Domain;
    fn sample(&self, rng: &mut R) -> Self::Domain;
}
