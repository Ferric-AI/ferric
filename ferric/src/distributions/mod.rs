// Copyright 2022 The Ferric AI Project Developers
mod bernoulli;
mod distribution;
mod poisson;

// Re-exports
pub use self::bernoulli::Bernoulli;
pub use self::distribution::Distribution;
pub use self::poisson::Poisson;
