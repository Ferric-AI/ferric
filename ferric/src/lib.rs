// Copyright 2022 The Ferric AI Project Developers

// re-export make_model from the ferric-macros crate
pub use ferric_macros::make_model;

// Public modules
pub mod core;
pub mod distributions;
pub use self::core::FeOption; // re-export FeOption
