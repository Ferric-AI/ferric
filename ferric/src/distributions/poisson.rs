// Copyright 2022 The Ferric AI Project Developers
use rand_distr::Distribution as Distribution2;
use rand_distr::Poisson as Poisson2;

use crate::distributions::Distribution;
use rand::Rng;

pub struct Poisson {
    /// Probability of success.
    rate: f64,
}

impl Poisson {
    pub fn new(rate: f64) -> Result<Poisson, String> {
        if rate <= 0f64 {
            Err(format! {"Poisson: illegal rate `{}` should be greater than 0", rate})
        } else {
            Ok(Poisson { rate })
        }
    }
}

impl<R: Rng + ?Sized> Distribution<R> for Poisson {
    type Domain = u64;
    fn sample(&self, rng: &mut R) -> u64 {
        Poisson2::new(self.rate).unwrap().sample(rng) as u64
    }
}

impl std::fmt::Display for Poisson {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Poisson {{ rate = {} }}", self.rate)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::thread_rng;

    #[test]
    fn poisson_sample() {
        let mut rng = thread_rng();
        let rate = 2.7f64;
        let dist = Poisson::new(rate).unwrap();
        println!("dist = {}", dist);
        let mut total = 0u64;
        let trials = 10000;
        for _ in 0..trials {
            total += dist.sample(&mut rng);
        }
        let mean = (total as f64) / (trials as f64);
        let err = 5.0 * (rate / (trials as f64)).sqrt();
        println!("empirical mean is {} 5sigma error is {}", mean, err);
        assert!((mean - 2.7).abs() < err);
    }

    #[test]
    #[should_panic]
    fn poisson_too_low() {
        let _dist = Poisson::new(-0.01).unwrap();
    }
}
