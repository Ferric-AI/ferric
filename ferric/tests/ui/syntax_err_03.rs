// Copyright 2022 The Ferric AI Project Developers
use ferric::make_model;

make_model! {
    mod grass;
    use ferric::distributions::Bernoulli;

    letu rain : bool ~ Bernoulli::new( 0.2 );

    let sprinkler : bool ~
        if rain {
            Bernoulli::new( 0.01 )
        } else {
            Bernoulli::new( 0.4 )
        };

    let grass_wet : bool ~ Bernoulli::new(
        if sprinkler && rain { 0.99 }
        else if sprinkler && !rain { 0.9 }
        else if !sprinkler && rain { 0.8 }
        else { 0.0 }
    );

    observe grass_wet;
    query rain;
    query sprinkler;
}

fn main() {
    let _model = grass::Model { grass_wet: true };
}
