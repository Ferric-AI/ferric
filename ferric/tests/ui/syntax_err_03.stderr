error: expected let | use | observe | query
 --> tests/ui/syntax_err_03.rs:8:5
  |
8 |     letu rain : bool ~ Bernoulli::new( 0.2 );
  |     ^^^^

error[E0433]: failed to resolve: use of undeclared crate or module `grass`
  --> tests/ui/syntax_err_03.rs:30:18
   |
30 |     let _model = grass::Model { grass_wet: true };
   |                  ^^^^^ use of undeclared crate or module `grass`
